python/sanlock.c

```diff
@@ -918,50 +915,103 @@ py_inq_lockspace(PyObject *self __unused, PyObject *args, PyObject *keywds)
     return NULL;
 }
 
+static int
+rem_lockspace_parse_args(
+        PyObject* args,
+        PyObject* keywds,
+        struct sanlk_lockspace* ls,
+        int* flags)
+{
+    int rv = 0, unused = 0;
+    PyObject *lockspace = NULL, *path = NULL;
+    /* initialize parsed args */
+    memset(ls, 0, sizeof(struct sanlk_lockspace));
+    *flags = 0;
+
+#if PY_MAJOR_VERSION == 2
+    /* parse arg for python 2 api */
+    int async = -1, wait = -1;
+    static char* kwlist[] = {"lockspace", "host_id", "path", "offset",
+                                "async", "unused", "wait", NULL};
+
+    if(!PyArg_ParseTupleAndKeywords(args, keywds, "O&kO&|kiii", kwlist,
+        convert_to_pybytes, &lockspace, &(ls->host_id), pypath_converter, &path,
+        &(ls->host_id_disk.offset), &async, &unused, &wait)) {
+        goto finally;
+    }
+    if (wait != -1 && async != -1) {
+        PyErr_Format(PyExc_RuntimeError, "async and wait flags must be mutually exclusive");
+        goto finally;
+    }
+    if (wait == -1) {
+        wait = 1;
+        if (async != -1)
+            wait = !async;
+    }
+#else
+    /* parse args for python 3 api */
+    int wait = 1;
+    static char* kwlist[] = {"lockspace", "host_id", "path", "offset",
+                                "unused", "wait", NULL};
+    if(!PyArg_ParseTupleAndKeywords(args, keywds, "O&kO&|kii", kwlist,
+        convert_to_pybytes, &lockspace, &(ls->host_id), pypath_converter, &path,
+        &(ls->host_id_disk.offset), &unused, &wait)) {
+        goto finally;
+    }
+#endif
+    /* prepare sanlock names */
+    strncpy(ls->name, PyBytes_AsString(lockspace), SANLK_NAME_LEN);
+    strncpy(ls->host_id_disk.path, PyBytes_AsString(path), SANLK_PATH_LEN - 1);
+
+    /* prepare sanlock_rem_lockspace flags */
+    if (!wait) {
+        *flags |= SANLK_REM_ASYNC;
+    }
+
+    if (unused) {
+        *flags |= SANLK_REM_UNUSED;
+    }
+    rv = 1;
```

Comment:
Can we have smaller and simpler functions that only do the argument parsing, and call common
function to do the actual action?

For example:

```
def rem_lockspace_py2(async=None, wait=None):
    parse args for py2
    rem_lockspace_internal(wait)

def rem_lockspace_py3(wait):
    parse args for py3
    rem_lockspace_internal(wait)
```
We can register the the py2 version when building py2 extension, and the py3 version when building py3 version.

Reply:
we can, but it will amount to same amount of code. and i dont see any difference.